// @sap/approuter/lib/backend-response/cookie-handler.js
'use strict';

var tough = require('tough-cookie');
var sessionExt = require('../utils/session-ext');
var cookieUtils = require('../utils/cookie-utils');

exports.processCookies = function (setCookieHeader) {
  let sessionCookies = [];
  let nonSessionCookies = [];
  if (setCookieHeader) {
    for (let i = 0; i < setCookieHeader.length; ++i) {
      let oCookie = tough.Cookie.parse(setCookieHeader[i]);
      if (oCookie && isSessionCookie(oCookie)) {
        sessionCookies.push(oCookie);
      } else {
        nonSessionCookies.push(setCookieHeader[i]);
      }
    }
  }
  return {
    sessionCookies: sessionCookies,
    nonSessionCookies: nonSessionCookies
  };
};

exports.storeSessionCookies = function(sessionCookies, req, res) {
  const isService2Approuter = req.headers['x-approuter-authorization'];
  if (sessionCookies.length === 0) {
    return;
  }
  let logger = req.loggingContext.getLogger('/StoreSessionCookies');
  if (isService2Approuter && res) {
    logger.info('Sending encrypted backend session cookies to client');
    const backendName = cookieUtils.getBackendName(req);
    let mergedCookies = mergeCookies(req.headers['cookie'],sessionCookies,backendName,logger);
    cookieUtils.setCookie(res, cookieUtils.encryptCookies(mergedCookies, backendName));
    return;
  }
  sessionExt.update(req.session, function (session) {
    let cookieJar = session._cookieJar ? tough.CookieJar.deserializeSync(session._cookieJar) : new tough.CookieJar();
    sessionCookies.forEach(function (oCookie) {
      try {
        cookieJar.setCookieSync(oCookie, req.internalUrl.href);
      } catch (err) {
        logger.warning('Could not set session cookie: ', err.message);
      }
    });
    session._cookieJar = cookieJar.serializeSync();
  });
};

function isSessionCookie(oCookie) {
  return (!oCookie.expires || oCookie.expires === 'Infinity') && (!oCookie.maxAge || oCookie.maxAge === 'Infinity');
}

function mergeCookies(clientCookies,backendCookies, backendName, logger){
  if (!clientCookies || process.env.SERVICE_2_APPROUTER_DISABLE_COOKIE_MERGE){
    return backendCookies;
  }
  let decryptedClientCookies = cookieUtils.decryptCookies(clientCookies, backendName).split(';');
  decryptedClientCookies.forEach(function(clientCookie){
    let [clientCookieKey, clientCookieValue] = clientCookie.split('=');
    let backendCookieIndex = null;
    for (let i = 0; i < backendCookies.length; i++) {
      if (backendCookies[i].key === clientCookieKey){
        backendCookieIndex = i;
        break;
      }
    }
    let shouldDelete = shouldDeleteBackendCookie(backendCookies, backendCookieIndex,clientCookieValue, logger);
    if (backendCookieIndex === null && !shouldDelete){
      backendCookies.push({
        key: clientCookieKey,
        value: clientCookieValue,
        path: '/',
        creation: new Date()
      });
    } else if (shouldDelete){
      backendCookies.splice(backendCookieIndex);
    }
  });
  return backendCookies;
}

function shouldDeleteBackendCookie(backendCookies, backendCookieIndex,clientCookieValue, logger){
  if (!backendCookies || backendCookies.length === 0 || backendCookieIndex === null){
    return false;
  }
  if (backendCookies[backendCookieIndex].maxAge === 0 || backendCookies[backendCookieIndex].value === clientCookieValue){
    logger.info('Backend cookie ' + backendCookies[backendCookieIndex].key + 'with value ' + backendCookies[backendCookieIndex].value + ' will be deleted');
    return true;
  } else {
    return false;
  }
}
